Core/src/SpacePointFormation2/StripSpacePointBuilder.cpp

```diff
@@ -0,0 +1,308 @@
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "Acts/SpacePointFormation2/StripSpacePointBuilder.hpp"
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/SpacePointFormation2/PixelSpacePointBuilder.hpp"
+#include "Acts/Utilities/MathHelpers.hpp"
+#include "Acts/Utilities/VectorHelpers.hpp"
+
+namespace Acts {
+
+std::optional<double> StripSpacePointBuilder::computeClusterPairDistance(
+    const Vector3& globalCluster1, const Vector3& globalCluster2,
+    const ClusterPairingOptions& options) {
+  // Check if measurements are close enough to each other
+  if ((globalCluster1 - globalCluster2).norm() > options.maxDistance) {
+    return std::nullopt;
+  }
+
+  const Vector3 vertexToCluster1 = globalCluster1 - options.vertex;
+  const Vector3 vertexToCluster2 = globalCluster2 - options.vertex;
+
+  // Calculate the angles of the vectors
+  const double phi1 = VectorHelpers::phi(vertexToCluster1);
+  const double theta1 = VectorHelpers::theta(vertexToCluster1);
+  const double phi2 = VectorHelpers::phi(vertexToCluster2);
+  const double theta2 = VectorHelpers::theta(vertexToCluster2);
+
+  // Calculate the squared difference between the theta angles
+  const double diffTheta2 = square(theta1 - theta2);
+  if (diffTheta2 > options.maxAngleTheta2) {
+    return std::nullopt;
+  }
+
+  // Calculate the squared difference between the phi angles
+  const double diffPhi2 = square(phi1 - phi2);
+  if (diffPhi2 > options.maxAnglePhi2) {
+    return std::nullopt;
+  }
+
+  // Return the squared distance between both vector
+  return diffTheta2 + diffPhi2;
+}
+
+bool StripSpacePointBuilder::computeCosmicSpacePoint(
+    const StripEnds& stripEnds1, const StripEnds& stripEnds2,
+    Vector3& spacePoint) {
+  // This approach assumes that no vertex is available. This option aims to
+  // approximate the space points from cosmic data.
+  // The underlying assumption is that the best point is given by the closest
+  // distance between both lines describing the SDEs. The point x on the first
+  // SDE is parametrized as a + lambda0 * q with the top end a of the strip and
+  // the vector q = a - (bottom end of the strip). An analogous parametrization
+  // is performed of the second SDE with y = c  + lambda1 * r. x get resolved by
+  // resolving lambda0 from the condition that |x-y| is the shortest distance
+  // between two skew lines.
+
+  const Vector3 firstBtmToTop = stripEnds1.top - stripEnds1.bottom;
+  const Vector3 secondBtmToTop = stripEnds2.top - stripEnds2.bottom;
+
+  const Vector3 ac = stripEnds2.top - stripEnds1.top;
+  const double qr = firstBtmToTop.dot(secondBtmToTop);
+  const double denom = firstBtmToTop.dot(firstBtmToTop) - qr * qr;
+  // Check for numerical stability
+  if (std::abs(denom) < 1e-6) {
```

Comment:
Hmm magic value, do we have some suitable `constexpr` threshold for such things in the library? If not, maybe a local named constant would be nice.